centos9/ansible11

jturmelle opened the pull request:
This pull request is to create a new CentOS Stream 9 installation with ansible-11.  This requires python3.12, so I've configured this to create a global python environment in /opt/datalib_venv3.12.  I considered putting it in /usr/local/datalib but didn't think it appropriate for some reason.

The documentation for this PR is here: [https://github.com/iridl/iridl.github.io/pull/14](https://github.com/iridl/iridl.github.io/pull/14).

I've tested it as a new install, but haven't tested it as an upgrade to pre-install locations (NiMet).

aaron-kaplan commented:
NiMet is unique, so we can handle it in an _ad hoc_ way. The thing that needs to be well-tested and well-documented is upgrading a CentOS 7 system. Ideally we get that smooth enough that the regional centers can do some of the work for us.

aaron-kaplan commented on example/ansible.cfg:
```diff
@@ -1,3 +1,4 @@
 [defaults]
-collections_paths = .
+ansible_python_interpreter = /opt/datalib_venv3.12/bin/python3
```
I thought we decided to use the CentOS-packaged ansible, which I would presume uses the system python.

jturmelle replied:
The ansible version installed with the system uses ansible 2.14, but the community.docker collection requires ansible 2.15.  Additionally, we'd need to pin the python-requests to a previous version in order for docker pull to work correctly. So I decided to go with a supported version of ansible that uses the latest community.docker because who knows what other bugs we'll encounter.  I could probably get everything to work properly with ansible 2.14, but I thought it would be better to have the most recent packages.

aaron-kaplan replied:
OK, that's too bad.

Having to remember to activate the virtualenv before running ansible-playbook or ansible-galaxy is not great. For that matter, the ansible-playbook command itself is too long to remember. Should we make a little `update-config` script that runs ansible-playbook with all the right arguments, in the right virtualenv?

I really think ansible may be a net negative for partner configs. I have half a mind to spend a day trying to rewrite the whole thing in bash and see how far I can get.

jturmelle replied:
That sounds a lot easier than rewriting everything in bash!

aaron-kaplan commented on roles/iridl/tasks/main.yaml:
```diff
@@ -242,7 +228,7 @@
     path: /root/.ssh
     state: directory
-    mode: 0700
+    mode: '0700'
```
Out of curiosity, was this necessary, or just a stylistic preference?

aaron-kaplan commented on roles/iridl/tasks/main.yaml:
```diff
@@ -404,42 +390,36 @@
 
 - debug:
     msg: "{{docker_compose_result.actions}}"
-  when: python_maproom_repo is defined
```
What was wrong with this? The [docs](https://docs.ansible.com/ansible/latest/playbook_guide/playbooks_conditionals.html) still use this syntax.

jturmelle replied:
At one point I had added a second conditional for some reason and never changed the syntax back. I've changed it back for consistency.

jturmelle commented on roles/iridl/templates/docker_daemon.j2:
```diff
@@ -0,0 +1,8 @@
+{
+    "log-driver": "local",
```
OK, I didn't think of it.

jturmelle commented on roles/iridl/handlers/main.yaml:
```diff
@@ -11,30 +11,21 @@
-    # Unfortunately, ansible's docker_compose module overrides the
-    # stop_grace_period specified in docker-compose.yaml, so we also
-    # have to configure it here.
-    # See https://github.com/ansible-collections/community.general/issues/293
```
I haven't tested it, but will see if I can confirm.

aaron-kaplan replied:
This is hard to test, and not worth spending a lot of time on. Squid restarts are rare, and partner DLs don't get a lot of traffic, so the number of dropped connections will be small even if we get this wrong.

jturmelle reviewed (commented):
Regarding upgrading from CentOS7, a full re-install of the OS is required, so a simple upgrade path is not possible. What I could probably do is provide some guidance on backing up and recovering their data in the new OS.

aaron-kaplan commented:
> Regarding upgrading from CentOS7, a full re-install of the OS is required, so a simple upgrade path is not possible. What I could probably do is provide some guidance on backing up and recovering their data in the new OS.

Yes, that's what I mean. We can't afford to hold the hands of all the former partners who need to upgrade but no longer have paying projects. We need to point them to documentation on the steps to take.

aaron-kaplan commented:
Please update the Vagrant provisioner. It still uses the original CentOS 7 installation method.

Also, scanning my previous comments, the following remain to be addressed:
- Provide a script that runs ansible with the right arguments, in the right virtual environment. Should support the --build-classic option.
- Test that when the handler restarts squid, it gives it 30 seconds to drain connections before killing it. This may actually be hard to test in a test environment, unless you can figure out how to make a long-running request. Worst case, just remember to test it as soon as you've deployed it to production, because production squid always has lots of open connections to drain.
- Document how to upgrade from CentOS 7. That's not a blocker for this PR.

jturmelle commented:
I misunderstood and thought the upgrade documentation was a requirement.  So I spent a lot of time working on the CentOS7 upgrade installation instructions and generally working on the dldocs.iri pages.  So if that's not a blocker, I'll finish up my updates there and push them once I test out the oauth git authentication and see if I can include that.

I didn't even notice the Vagrant provisioner, and the comment regarding testing squid is useful.  I can get these done by tomorrow.

aaron-kaplan commented:
The upgrade doc is critical--nobody can run this new version of the playbook until they upgrade. Just saying this PR seems close to the finish line and I don't see any harm in merging it first.

aaron-kaplan commented on tests/Vagrantfile:
```diff
@@ -1,7 +1,7 @@
 # vi: set ft=ruby :
 
-$bootstrap = <<-EOF
+$centos7_bootstrap = <<-EOF
```
We don't need to continue to support centos7 on master. We said that if we need to make any updates to centos7, we'll make a separate branch for that.

jturmelle replied:
I thought it might be useful if we needed to do testing, but you're right.

jturmelle commented on roles/iridl/templates/docker-compose.yaml.j2:
```diff
@@ -18,7 +18,7 @@ services:
     # latter case, ansible overrides this timeout, so we have to
     # specify it redundantly in the "restart squid" handler.
-    stop_grace_period: "{{squid_shutdown_lifetime + 10}}s"
+    stop_grace_period: {{squid_shutdown_lifetime + 10}}s
```
When I was testing I was afraid it wasn't working, looked at the ansible documentation and there are no quotes in the documentation.  

I still haven't been able to test because I can't figure out how to run a long-running process with the test maprooms.

jturmelle commented on tests/Vagrantfile:
```diff
@@ -16,20 +16,50 @@ echo Rebooting...
+  
+  config.vm.define "dlserver9", primary: true do |x|
+    x.vm.box = "centos/stream9"
+    x.vm.disk :disk, size: '20GB', primary: true
```
default VM size is 10GB.  The new system uses about 15GB.

aaron-kaplan commented on example/install.bash:
```diff
@@ -0,0 +1,82 @@
+
+# Check for git changes
+echo "Checking for git changes"
+git fetch
```
I'm sympathetic to the argument that we should be able to run the playbook when the internet is unavailable.

On the other hand, if I make changes locally when there are changes upstream that I haven't pulled, then I screwed up and must not proceed until I fix it. Also, if fetch fails not because the internet is down but because  I haven't configured my git credentials correctly, I should really fix that before proceeding.

Furthermore, if the script forges ahead and runs ansible-playbook after printing this warning, nobody will ever see the warning, because ansible-playbook produces pages and pages of output, so the warning will quickly scroll out of view.

I take back what I said about preferring pull to fetch. If pull --ff-only fails, it leaves you in the same state as if you fetched but didn't pull, so it's not really any better.

So here's what I propose: offer a `--no-fetch` option. If `--no-fetch` isn't provided, then do a fetch, and abort if it fails. Then in either case, check whether origin/master and master point to the same commit, and abort if not.

aaron-kaplan commented on example/install.bash:
```diff
@@ -0,0 +1,80 @@
+echo "For more information, review https://dldocs.iri.columbia.edu/admin/maintenance.html"
+echo ""
+
+if [[ "${CHECK}" == "" ]]; then
```
BTW you broke this by setting `CHECK=-diff` at the top.

aaron-kaplan commented on example/install.bash:
```diff
@@ -0,0 +1,80 @@
+##
+VENV=/opt/datalib_venv3.12
+BUILD=""
+CHECK="-diff"
```
Two hyphens, not one.

But --diff really has nothing to do with --check, so I wouldn't put it in the CHECK variable. Just hard-code --diff in the command in line 80.

aaron-kaplan commented on example/install.bash:
```diff
@@ -0,0 +1,80 @@
+
+if [ $# -gt 0 ]; then
+  case $1 in
+    --build|-build|build)
```
The convention is two hyphens before a long option, one hyphen before a single-letter option, and no hyphens for a positional argument. Don't accept all three, only `--build`. Same comment for `--check` below.

aaron-kaplan commented on example/install.bash:
```diff
@@ -0,0 +1,80 @@
+if [[ "${CHECK}" == "" ]]; then
+  echo "You are running the command in check mode, no changes will be made."
+else
+  echo "You are about to make install your changes to the Data Library installation."
```
"make install" --> "apply".

But I would still prefer the `--no-fetch` option (in an earlier comment of mine). It's too easy to say "yes, I want to continue" without noticing that you've made a critical mistake (failed to merge upstream). If we *know* that upstream has diverged, we should not run the playbook under any circumstances. Not even if the user says to. That has to be fixed first. The only case where I'm ok with proceeding is if we fail to fetch, so we don't know whether upstream has diverged or not.

jturmelle replied:
I think I've applied what you're thinking.

aaron-kaplan commented on example/install.bash:
```diff
@@ -0,0 +1,82 @@
+#!/bin/bash
```
apply?

jturmelle replied:
I prefer install to apply.  Installing an upgrade is still valid while apply doesn't seem distinctive enough.  What if we  use the "apb" convention we use internally?

aaron-kaplan replied:
apb just seems a bit cryptic. run-ansible?

jturmelle replied:
perfect.  I've changed it.

jturmelle commented on roles/iridl/templates/docker-compose.yaml.j2:
```diff
@@ -18,7 +18,7 @@ services:
     # when we use ansible to stop or restart the service. In the
     # latter case, ansible overrides this timeout, so we have to
     # specify it redundantly in the "restart squid" handler.
```
I had thought to leave it to remind us it still needs to be tested, but I think once we test it internally for our DL, we can apply that knowledge to this.

aaron-kaplan commented on example/install.bash:
```diff
@@ -0,0 +1,114 @@
+gitstatus="$(git status)"
+echo ""
+
+behind="branch is behind \'origin"
```
Searching for English strings in the output is very brittle. I know that on the Niger server, for example, they use a Francophone locale, so errors messages are in French.

Git has "plumbing" and "porcelain" commands. Porcelain is for interactive use by humans; plumbing is for use by scripts. Plumbing provides a stable and locale-independent API.

jturmelle replied:
I tried porcelain, but it didn't always return the necessary output.  I didn't know about plumbing.  I'll look at that.

aaron-kaplan replied:
I didn't know `git status` had a  `--porcelain` option, but that's the plumbing I was referring to. (Confusing name. It's called `--porcelain` because it's for use by porcelain scripts.)

`git status --porcelain=2 --branch` shows some extra information, but I'm not sure it's useful. Maybe git status isn't the right command for this.

jturmelle replied:
Yes, _git status_ isn't the right command for this.  I tried git status --porcelain,  but it wasn't providing enough information when a branch was behind the master. I'm looking at _git diff_, which I think is the correct way to do it.

aaron-kaplan commented:
Checking for unpulled commits isn't the only thing you might want to check for. Are we on a non-default branch? Do we have uncommitted changes? Are there untracked files? Maybe it's a fool's errand to try to catch every possible violation.

Igor and I went through this same thought process when writing a container image release script. In the end we gave up on checking the status of the local git repo, and instead wrote a script that pulls down a fresh clone from bitbucket every time you run it. In that scenario, the admin wouldn't even need to run git on the server, so no configuring an ssh key. Just make your changes in the bitbucket web UI, then run `run_ansible` on the server. A bit of a slow edit-test cycle, but it might be better than the status quo?

Creating the repo in the first place might be a bit tricky, but we could just do that for them ahead of time.

aaron-kaplan commented on tests/run_tests:
```diff
@@ -142,6 +139,7 @@ _playbook() {
             return 1
             ;;
     esac
+    export ANSIBLE_CONFIG=../example/ansible.cfg
```
Why wasn't this needed before?

jturmelle commented:
> Checking for unpulled commits isn't the only thing you might want to check for. Are we on a non-default branch? Do we have uncommitted changes? Are there untracked files? Maybe it's a fool's errand to try to catch every possible violation.
>
I think warning them if they have uncommitted changes or if they are behind the master is all we should/can do.  And now even that is becoming problematic.  I think this script is a waste of time and we should go back to the original ansible_playbook instructions and simply provide them with some guidance in the documentation on how to use git to check their status.  This also gives them more leeway on how they interact with git, for those who are familiar with it.

I advocate for providing them with a one-page cheat sheet they can use to install changes, check their git status and whatever else might come up.  We update this cheat sheet as we run into problems.

We can also give them a script they _can_ use to check their git status and warn them of possible issues: such as they're not on the main branch, they have uncommitted changes, or they are not up to date with the master; and let them use that information as they see fit.

I can still work on a script like that, but I don't think it should stop us from moving forward.

jturmelle commented:
ansible was previously installed in the default python interpreter.  

> On Mar 17, 2025, at 4:49 PM, Aaron Kaplan ***@***.***> wrote:
> 
> 
> @aaron-kaplan commented on this pull request.
> 
> In tests/run_tests <https://github.com/iridl/iridl-ansible/pull/18#discussion_r1999650668>:
> 
> > @@ -142,6 +139,7 @@ _playbook() {
>              return 1
>              ;;
>      esac
> +    export ANSIBLE_CONFIG=../example/ansible.cfg
> Why wasn't this needed before?
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/iridl/iridl-ansible/pull/18#pullrequestreview-2692047635>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABJRNUDLY5SVP5IG2BNWVPD2U4YOPAVCNFSM6AAAAABYWXUHC2VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDMOJSGA2DONRTGU>.
> You are receiving this because you were assigned.
>

aaron-kaplan commented:
> I think this script is a waste of time and we should go back to the original ansible_playbook instructions

The original purpose of this script wasn't to enforce git safety checks, it was to save people from having to remember and type a very long command that got even longer now that you have to activate a venv. I agree with dropping the git checks, but not with eliminating the script.

jturmelle commented on example/ansible.cfg:
```diff
@@ -1,3 +1,4 @@
 [defaults]
-collections_paths = .
+interpreter_python = /opt/datalib_venv3.12/bin/python3
```
I've moved it to inventory.cfg and placed it on the dlservers line so it won't be a global.

jturmelle commented on example/run_ansible:
```diff
@@ -0,0 +1,115 @@
+#!/bin/bash
```
Done.  Let me know if you just want me to rename it in this PR.

jturmelle commented on example/run_ansible:
```diff
@@ -0,0 +1,115 @@
+  shift
+done
+
+# Check if the virtual environment exists
```
They need the virtual environment to already be installed so they can run ansible-galaxy to install the iridl collection so they can get this script.

Alternatively, we could make this script available as a download from dldocs.iri.columbia.edu as a bootstrap.

aaron-kaplan replied:
Ugh. Bootstrapping ansible.

A `curl | bash` setup might be handy. But that would be a separate script, not this script, I think.

aaron-kaplan commented on example/git_check.bash:
```diff
@@ -0,0 +1,125 @@
+#!/bin/bash
```
You removed this file in the previous commit and then added it back in the latest one.

jturmelle replied:
I renamed it so I could remove it and keep it to work on later. Then installed the new copy accidentally.

aaron-kaplan commented on example/run-ansible:
```diff
@@ -0,0 +1,67 @@
+    --check)
+      CHECK="--check"
+      ;;
+    --no-fetch)
```
We don't need this option anymore.

aaron-kaplan commented:
`./run_tests full` provisions the VM, restarts it, then fails to run ansible. I ran it with `bash -x` and got the following output

```
+ vagrant ssh -- exec bash -l -c ''\''/vagrant/tests/./run_tests --mode aux full'\'''
--mode
full
/vagrant/tests/./run_tests: line 96: ansible-galaxy: command not found
```
`run_tests` (argh! an underscore!) should run `run-ansible` rather than `ansible-playbook` now.

If you can't run Vagrant on your mac, can you run it on mako?

jturmelle commented:
> `./run_tests full` provisions the VM, restarts it, then fails to run ansible. I ran it with `bash -x` and got the following output
> 
> ```
> + vagrant ssh -- exec bash -l -c ''\''/vagrant/tests/./run_tests --mode aux full'\'''
> --mode
> full
> /vagrant/tests/./run_tests: line 96: ansible-galaxy: command not found
> ```
> 
> `run_tests` (argh! an underscore!) should run `run-ansible` rather than `ansible-playbook` now.
> 
> If you can't run Vagrant on your mac, can you run it on mako?

I was a little confused about how to run run_tests. I do have a working vagrant on a spare laptop and will try it again.  I never tried it in guest mode, although it worked in full mode (until it got to conda, which I've not been able to debug yet).

aaron-kaplan commented:
Guest mode is the default because that's how partners run it: the Vagrant guest is both the ansible control node and the target node. In host mode, the vagrant host is the control node and the guest is the target.

jturmelle commented:
I made some changes to the tests to work properly with run-ansible, and increased memory requirements. I don't think I can use run-ansible in host mode. Tested in Vagrant, but I still wasn't able to get the ansible_test portion to work because I think it requires sudo privileges to access docker.  I can't get it to work and don't know why.

aaron-kaplan commented on tests/run-tests:
```diff
@@ -130,24 +119,38 @@ _up() {
+          -e @vars-iri.yaml \
+          -e @../example/secrets.yaml \
+          ../example/playbook.yaml ${extraflags}
+      ;;
```
Let's simplify by removing host mode. It exists because originally that's all there was; then we started thinking about running ansible on the server itself for real installs, so I added guest mode so that I could test it that way. Now that we've settled on that approach and are using it everywhere, I see no reason to carry the extra complexity anymore.

I think that means `tests/inventory.cfg` can be deleted, as it was only used in host mode.

aaron-kaplan replied:
Oh, another reason host mode is still around is that under centos 7 the guest had ansible 2.9, and ansible_test didn't work there. That's no longer a concern, so you should be able to get rid of the switch on `$mode` in line 171-178 now, and allow ansible_test in guest mode.

That reminded me to run ansible_test, and it fails. I think it's false positives, will look into how to suppress them.

aaron-kaplan commented on tests/run-tests:
```diff
@@ -130,24 +119,38 @@ _up() {
+    aux)
+      mkdir ${builddir}/dlconfig
+      cp ../example/* ${builddir}/dlconfig
+      cp ../example/secrets.yaml ${builddir}
```
This is kind of crufty, but I can't think of a better way at the moment. I think I want to stop using secrets.yaml, in favor of a certificate generated on the server and installed in bitbucket via the HTTP API. With secrets.yaml eliminated, we could probably do without this copy. But that's for another PR.

aaron-kaplan replied:
Hmm. Leaving secrets.yaml in the project directory and adding it to .gitignore is a tempting solution. But I generally assume that a git repo with no uncommitted changes can be wiped out and re-cloned without disrupting anything; it would be annoying if someone did that and then found they no longer had a copy of their secrets file. (Annoying but not fatal--you would just generate and install new keys.)

aaron-kaplan replied:
I think moving secrets.yaml would be more trouble than it's worth, as many partners already have it set up this way.

aaron-kaplan reviewed (commented):
Yes ansible needs sudo privileges, but the vagrant box allows sudo with no password. You're being prompted for the password by ansible (not by sudo) because `run-tests` hard-codes the option `--ask-become-pass`. We'll have to add an option to `run-tests` to tell it not to use that option.

BTW there are still two different keys in secrets.yaml. Weren't you going to collapse them into one? That can be deferred to another PR if you want.

jturmelle reviewed (commented):
Regarding ansible-test, it wasn't the password option that was the problem, but I believe it was an issue accessing the GitHub repository as sudo.  I can't remember specifically, but it seemed like an impossible combination.  When I remove host mode, I'll rerun and figure out exactly what the problem was.

I didn't change secrets.yaml, mainly because of the comment you made regarding another country using their own git repository accounts so it's not a bad idea to be flexible.  I updated the documentation to say they can use the same key in both locations though.  Alternatively, I could change the template to use the bitbucket key if the python_maproom key isn't there.

aaron-kaplan commented on tests/run-tests:
```diff
@@ -184,6 +187,7 @@ while (( $# )); do
                     $1 || fail
                     ;;
                 aux)
+                    . /opt/datalib_venv3.12/bin/activate
```
Ah, forgot about ansible-galaxy.

I was thinking maybe we symlink ansible-playbook and ansible-galaxy into /usr/local/bin. (With a virtualenv, unlike with a conda environment, you don't necessarily have to activate it. It's sufficient to run commands from the environment using their full path.) Activating the environment in /etc/profile.d could be another option, but I think the symlink idea has less possible impact on other things.

jturmelle commented:
After spending hours on Friday trying to debug a problem with git access_keys, I realize today the problem must have originated at bitbucket because now it's all working with no changes.  Bitbucket had a status report that service was degraded, but it didn't say anything about ssh access keys, so I thought it was me.  Just a note if you start running into unexplained access denied errors!

I committed a couple minor changes to the run-tests script, and there is still an issue with ansible-test.  ansible-test needs root privileges to run, and needs access to the virtual environment. It got around this before because it was only available in host mode, which we're not doing any more. I'm not sure the best way to resolve this except making a secondary script for ansible-test, but this script already feels kludgy enough. Mainly because I tried to adapt your script instead of just rewriting it.  But we're here now, so let me know what you want me to do with it.  It does seem to work for everything except ansible-test right now.

aaron-kaplan commented on tests/vars-iri.yaml:
```diff
@@ -1,7 +0,0 @@
-# Variable overrides for testing at IRI.
-
-# Public NTP servers are blocked inside the IRI network, so
-# without this override the chronyc test fails when run at IRI.
```
Why did you delete this file? If you're having trouble with the crony tests, I think this is why. If you don't override the ntp server list, crony on the vagrant guest tries to contact outside NTP servers, which are blocked (I assume by the Lamont firewall).

aaron-kaplan commented on tests/run-tests:
```diff
@@ -0,0 +1,166 @@
+if [[ "${myname}" == "vagrant" ]]; then
+  guest=1
+else
+  _up
```
The old script called `_up` at the start of each of the other commands. Why did you move it here?

`./run-tests destroy full` is how I typically invoke the script. In this version, `_up` is called before `destroy`, so by the time we get to `full` the VM has already been provisioned and torn down, and there's no more VM to run the tests on.

aaron-kaplan reviewed (commented):
Is this working for you? I thought you were asking for a final review, but I've worked around at least three different problems and it's still not working for me, so maybe you were just asking for feedback on a draft?

I don't see anything wrong with the overall approach, if that's what you're asking.

I haven't found ansible-test to be tremendously valuable, sof we have to lose it, I won't be too sad. I don't understand your explanation of the problem, though. First of all, why does ansible-test need root; and second, ansible-playbook already needs root anyway, so why is it a problem that ansible-test does too?

jturmelle commented:
I've pushed some new commits that I believe have resolved everything.  I added a "hidden" flag to run-ansible --nopw (no password), which will use --become instead of --ask-become-pass.  I also added -v flag for use with run-ansible to pass to the ansible-playbook command.  I changed a couple files to use #!/usr/bin/env bash which was required by ansible-test to pass.

aaron-kaplan commented on tests/Vagrantfile:
```diff
@@ -1,35 +1,38 @@
+dnf install -y git python3.12
+python3.12 -m venv /opt/datalib_venv3.12
+/opt/datalib_venv3.12/bin/pip install --upgrade pip
+/opt/datalib_venv3.12/bin/pip install ansible==11.3.0 requests==2.32.3
```
We used to just pin `ansible`, it worked for a while, then it stopped working. For now we can work around it by also pinning `requests`, but that won't work forever either. Someday another package will make a backwards-incompatible release and installs will stop working again until we pin it. The only reproducible way to use pip is to freeze the entire dependency list (`pip freeze > requirements.txt`, then build the environment with `pip install -r requirements.txt`.

We can defer this to another PR, but I would like to address it eventually.

aaron-kaplan reviewed (commented):
I'm able to run it now.

The command I would like to run in CI is `./run-tests full idempotency`.  Currently `full`  succeeds but then `idempotency` fails because it's supposed to end with `changed=0` but in fact the task `add host keys to ssh_known_hosts for git repositories` makes a change. I tracked this down to `update_datalib`: each time I run that, the `ssh-rsa` keys for bitbucket and github reappear in `/root/.ssh/known_hosts`. To fix this, I suggest that instead of using `state: absent` we have the playbook either ignore the `ssh-rsa` key or set it to the new value (we originally added the `state: absent` to remove an old key that was revoked).

I find the `debug` task that follows `Update docker-compose state for services pulled from registries` excessively verbose. It produces about as much output as all the other tasks put together, which makes it harder to scan quickly through the output.

aaron-kaplan commented on roles/iridl/templates/generate-forecasts:
```diff
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/env bash
```
I would rather not make silly changes just to placate ansible-test. I don't know that it has ever found a true mistake yet, so if we have to do crap like this it might not be worth it.

That said, according to the [docs](https://docs.ansible.com/ansible-core/2.18/dev_guide/testing/sanity/shebang.html) `#!/bin/bash -eu` is one of the acceptable shebangs. This script already uses `set -e` on line 3, and `-u`  seems like a good idea in general, so let's go with that.

aaron-kaplan commented on roles/iridl/templates/upload-forecasts:
```diff
@@ -1,5 +1,4 @@
-#!/bin/bash
-
+#!/usr/bin/env bash
 set -e
```
`#!/bin/bash -eu`, see comment on generate-forecasts.

aaron-kaplan commented on example/run-ansible:
```diff
@@ -0,0 +1,67 @@
+#!/usr/bin/env bash
```
See comment on generate-forecasts. In this case the script doesn't explicitly use `set -e` already, but on a quick scan through it I don't think that would change the behavior.

jturmelle replied:
fixed

jturmelle reviewed (commented):
It appears to be working correctly now after resolving the /data/docker permissions.  I believe I've addressed all your comments.

aaron-kaplan commented on roles/iridl/templates/docker_daemon.j2:
```diff
@@ -0,0 +1,9 @@
+    "insecure-registries": ["{{docker_registry_mirror }}"],
+{% endif %}
+    "max-concurrent-downloads": {{docker_max_concurrent_downloads | default(3) | int}},
+    "data-root": "{{docker_dir}}"
```
We used to set `data-root` to a non-standard location, then we switched to leaving it at the default (`/var/lib/docker`) and making `/var/lib/docker` a symlink to `docker_dir`. Now you're adding this back. Seems confusing. Either remove this or remove the symlink.

jturmelle replied:
I don't see the confusion.  I added it back because this is where it's supposed to be set and I thought this might be part of the issue with the file permissions.  In my view, the link is there for people who aren't familiar with docker and may be looking for it there, ultimately limiting confusion.  I really feel this is the correct behavior, but if you feel strongly about it I'll go back to the symlink only.

aaron-kaplan replied:
If I make a separate volume for /var, and I decide that now I want to put the docker data root on that volume, I might remove the symlink and expect that to do it, but it won't.

jturmelle replied:
That might be true if we weren't using ansible, but we're setting a docker_dir variable specifically for this purpose.  Maybe we move the docker_dir variable to the playbook to make it more noticeable to the end user?

aaron-kaplan commented on example/run-ansible:
```diff
@@ -0,0 +1,73 @@
+{
+        echo ""
+        echo "Couldn't activate Python environment ${VENV}"
+        echo "Please review the documentation: https://dldocs.iri.columbia.edu/admin/server/centos9_stream.html#install_python"
```
As below, just link to the root of the admin docs for now.

jturmelle commented on roles/iridl/templates/docker_daemon.j2:
```diff
@@ -0,0 +1,9 @@
+    "registry-mirrors": ["{{docker_registry_mirror }}"],
+    "insecure-registries": ["{{docker_registry_mirror }}"],
+{% endif %}
+    "max-concurrent-downloads": {{docker_max_concurrent_downloads | default(3) | int}},
```
I don't see the confusion. It's just an error check if that variable isn't set, but I don't know why I changed it.

aaron-kaplan replied:
It's confusing because if I decide 3 was the wrong default and it should be 2 instead, I might come here and change the 3 to 2, and it will have no effect because the variable is always set. If it's not set in the playbook, then it falls back to the value from defaults/main.yaml.

jturmelle commented on tests/run-tests:
```diff
@@ -0,0 +1,170 @@
+                $1 || fail
+            else
+                _up || fail
+                vagrant ssh -- exec bash -l -c "'/vagrant/tests/$0 $1'" || fail
```
Not necessary because we cd to the directory just before that loop at line 148.

aaron-kaplan replied:
Line 148 doesn't change the value of `$0`. If I invoke this script as `/home/aaron/src/iridl-ansible/tests/run-tests`, then this will attempt to run `/vagrant/tests/home/aaron/src/iridl-ansible/tests/run-tests` on the vagrant box.

jturmelle replied:
Of course.  I don't know where my head was.

jturmelle reviewed (commented):
I'm pushing now so I can pull them on another host to test.

aaron-kaplan commented on example/run-ansible:
```diff
@@ -0,0 +1,73 @@
+if [ ! -f "../secrets.yaml" ]; then
+    echo "../secrets.yaml file doesn't exist.  Please read the documentation on deploy keys and Configuring your Data Library:"
+    echo "* https://dldocs.iri.columbia.edu/admin/git.html#deployment-keys"
+    echo "* https://dldocs.iri.columbia.edu/admin/installation.html#configure-your-data-library-repository"
```
The comment I thought I had made isn't there. I must have left it half-finished, or else I decided to come back to it after this PR was finished.

The structure of the docs on your [branch](https://github.com/iridl/iridl.github.io/pull/14) makes no sense to me. Installing the application software comes before installing the OS, which comes before purchasing the hardware? Let's finish the current PR first, with less-specific URLs, and then continue with that one.

jturmelle reviewed (commented):
I'll explain my thinking on the documentation in a dldocs PR once this is ready.

jturmelle commented:
I've committed and tested the changes you requested, including removing data-root for expediency, even though I'm not convinced :)
